components/core/src/event.rs

```diff
@@ -0,0 +1,139 @@
+
+// Sample event JSON payload (compatible with Segment.io)
+// {
+//   "type": "track",
+//   "event": "builder-project-create",
+//   "properties": {
+//     "clientid" : "0a5c0882-ade5-46cf-821d-8d3853cd0d41"
+//     "timestamp": "1479330000.13442404",
+//   }
+// }
+
+const CLIENT_ID_METAFILE: &'static str = "CLIENT_ID";
+
+#[derive(Debug, Clone)]
+pub struct Event {
+    name: String,
+    clientid: String,
+    timestamp: String,
+    properties: BTreeMap<String, String>,
+}
+
+impl Event {
+    pub fn new(name: &str, clientid: &str, timestamp: &str) -> Self {
+        let mut properties = BTreeMap::new();
+        properties.insert("timestamp".to_string(), timestamp.to_string());
+        properties.insert("clientid".to_string(), clientid.to_string());
+
+        Event {
+            name: name.to_string(),
+            clientid: clientid.to_string(),
+            timestamp: timestamp.to_string(),
+            properties: properties,
+        }
+    }
+}
+
+impl ToJson for Event {
+    fn to_json(&self) -> Json {
+        let mut p = BTreeMap::new();
+        for (key, value) in self.properties.iter() {
+            p.insert(key.to_string(), value.to_json());
+        }
+
+        let mut m = BTreeMap::new();
+        m.insert("type".to_string(), "track".to_string().to_json());
+        m.insert("event".to_string(), self.name.to_json());
+        m.insert("properties".to_string(), p.to_json());
+
+        Json::Object(m)
+    }
+}
+
+fn read_file(file_path: &Path) -> String {
+    let mut content = String::new();
+    let mut file = File::open(file_path).expect("Unable to open file");
+    file.read_to_string(&mut content).expect("Unable to read file");
+    content
+}
+
+fn write_file(parent_dir: &Path, file_path: &Path, content: &str) {
+    fs::create_dir_all(parent_dir).expect("Unable to create directory");
+    let mut file = File::create(&file_path).expect("Unable to create file");
+    file.write_all(content.as_bytes()).expect("Unable to write file");
+}
+
+fn timestamp() -> String {
+    let (secs, subsec_nanos) = match SystemTime::now().duration_since(UNIX_EPOCH) {
+        Ok(duration) => (duration.as_secs(), duration.subsec_nanos()),
+        Err(e) => {
+            debug!("Cannot generate system time: {}", e);
+            return "0.0".to_string();
+        }
+    };
+    format!("{}.{}", secs, subsec_nanos)
+}
+
+fn client_id() -> String {
+    let cache_dir = cache_analytics_path(None);
+    let file_path = cache_dir.join(CLIENT_ID_METAFILE);
+
+    if file_path.exists() {
+        read_file(&file_path)
+    } else {
+        let uuid = Uuid::new_v4().hyphenated().to_string();
+        write_file(&cache_dir, &file_path, &uuid);
+        uuid
+    }
+}
+
+pub fn record_event(name: &str) {
+    let timestamp: &str = &timestamp();
+    let clientid: &str = &client_id();
+    let event = Event::new(name, timestamp, clientid);
+
+    let cache_dir = cache_analytics_path(None);
```

Comment:
The `cache_analytics_path()` is for the client analytics but these are server side analytics. We want to store these instead in the `var` or `data` directory of the service. I think what we're looking for here is `core::fs::svc_var_path()`